### PR TITLE
Adding turnover policy controller

### PR DESCRIPTION
**Pull request by ch-mcabrera**
Code changes to add the turnover policy controller. 
- `turnoverPolicy model` 
Adding missing validation in the `turnover policy model`

- `ValidationMessages.properties` and `messages.properties`
Adding new error messages for the turnover policy validation.
- `ValidationMessage`
Adding new entry in the enumeration error mappings to cater for the latest `api's changes`. The api will trigger the same error, `invalid_input_length`, when string fields are either equals to the `min length` or `exceeds the max length`.  
The returned `invalid_input_length` from the api needs to map to `validation.length.maxExceeded`.  
- `Turnover Policy Service` 
Changes to cater for the latest `api's changes`. Validation added to check the `turnoverPolicy` model before sending it to the api. We should not send it to the api if it is `empty` or `contains multiple empty spaces`
- `turnoverPolicy` html and fragment `inlineYesNoRadioButtons` html
New fragment has been created for the `radio buttons` and for `the text box`. Not only will this be used by `turnoverPolicy.html` but also by other 3 accounting policy's screens. 
- JUnit test added for the controller
- `BasisOfPreparationController` and its JUnit test (`BasisOfPreparationControllerTests`) 
it had to change since it now leads to the `turnoverPolicy screen` instead of the `check your answer's screen`
- `ReviewScreenController` 
`PreviousController` had to change since it is now `turnoverPolicyController` instead of `BasisOfPreparationController`
- DisplayName's changes in `AccountingPoliciesTransformerImplTests.java` to make them more descriptive.

Resolves: SFA-631

Reliant on:
https://github.com/companieshouse/cdn.ch.gov.uk/pull/286

### CODE REVIEW

File: src/main/java/uk/gov/companieshouse/web/accounts/service/smallfull/impl/TurnoverPolicyServiceImpl.java [hunk 1]
```diff
@@ -39,14 +43,52 @@ public TurnoverPolicy getTurnOverPolicy(String transactionId, String companyAcco
...
+        return accountingPoliciesService.updateAccountingPoliciesApi(transactionId, companyAccountsId,
                 accountingPoliciesApi);
+    }
+
+    /**
+     * Validate the turnover policy details' length before calling the api. This string field should
+     * not be empty or contain empty characters.
+     *
+     * @param turnoverPolicy The turnover policy model that needs to be validated
+     * @return A list of validation errors. This will be empty fo
```
Comment (leahtaylor) [inline src/main/java/uk/gov/companieshouse/web/accounts/service/smallfull/impl/TurnoverPolicyServiceImpl.java]: is the @return of this javadoc missing the ending?

File: src/main/java/uk/gov/companieshouse/web/accounts/service/smallfull/impl/TurnoverPolicyServiceImpl.java [hunk 2]
```diff
@@ -39,14 +43,52 @@ public TurnoverPolicy getTurnOverPolicy(String transactionId, String companyAcco
...
+        return validationErrors;
+    }
+
+    /**
+     * Check if field contains invalid information: empty or contain empty characters.
+     *
+     * @param field the string field that needs to be assessed
+     * @return
+     */
+    private boolean isInValidStringField(String field) {
```
Comment (leahtaylor) [inline src/main/java/uk/gov/companieshouse/web/accounts/service/smallfull/impl/TurnoverPolicyServiceImpl.java]: should this be Invalid rather that In Valid?

Comment (leahtaylor) [reply src/main/java/uk/gov/companieshouse/web/accounts/service/smallfull/impl/TurnoverPolicyServiceImpl.java]: Actually - if this isjust checking if the field is empty should it just be named something like IsFieldEmpty?

File: src/main/java/uk/gov/companieshouse/web/accounts/service/smallfull/impl/TurnoverPolicyServiceImpl.java [hunk 3]
```diff
@@ -39,14 +43,52 @@ public TurnoverPolicy getTurnOverPolicy(String transactionId, String companyAcco
...
+        return StringUtils.trim(field).isEmpty();
     }
+
```
Comment (leahtaylor) [inline src/main/java/uk/gov/companieshouse/web/accounts/service/smallfull/impl/TurnoverPolicyServiceImpl.java]: empty line not required

File: src/main/resources/messages.properties [hunk 4]
```diff
@@ -74,4 +74,8 @@ validation.date.invalid.approval.date = The date cannot be on or before the end
 
 validation.characters.invalid.accounting_policies.basis_of_measurement_and_preparation = Enter valid characters
 validation.element.missing.accounting_policies.basis_of_measurement_and_preparation = Give details of how your accounts were prepared
-validation.length.maxExceeded.accounting_policies.basis_of_measurement_and_preparation = Basis of preparation must not exceed {0} characters
\ No newline at end of file
+validation.length.maxExceeded.accounting_policies.basis_of_measurement_and_preparation = Basis of preparation must not exceed {0} characters
+
+validation.characters.invalid.accounting_policies.turnover_policy = Enter valid characters
+validation.length.minInvalid.accounting_policies.turnover_policy = Give details of how your 'turnover policy'
```
Comment (leahtaylor) [inline src/main/resources/messages.properties]: is this error message complete?

File: src/main/resources/messages.properties [hunk 5]
```diff
@@ -74,4 +74,8 @@ validation.date.invalid.approval.date = The date cannot be on or before the end
...
+validation.length.invalidInputLength.accounting_policies.turnover_policy = Turnover Policy must not exceed {1} characters
```
Comment (leahtaylor) [inline src/main/resources/messages.properties]: Should policy have a lower case P?

File: src/test/java/uk/gov/companieshouse/web/accounts/controller/smallfull/TurnoverPolicyControllerTest.java [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+            .thenThrow(ServiceException.class);
+
+        this.mockMvc
+            .perform(get(TURNOVER_POLICY_PATH))
+            .andExpect(status().isOk())
+            .andExpect(view().name(ERROR_VIEW))
+            .andExpect(model().attributeExists(BACK_BUTTON_MODEL_ATTR))
+            .andExpect(model().attributeExists(TEMPLATE_NAME_MODEL_ATTR));
+    }
+
```
Comment (leahtaylor) [inline src/test/java/uk/gov/companieshouse/web/accounts/controller/smallfull/TurnoverPolicyControllerTest.java]: white space

File: src/test/java/uk/gov/companieshouse/web/accounts/controller/smallfull/TurnoverPolicyControllerTest.java [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+        this.mockMvc
+            .perform(createPostRequestWithParam(false))
+            .andExpect(status().isOk())
+            .andExpect(view().name(ERROR_VIEW))
+            .andExpect(model().attributeExists(BACK_BUTTON_MODEL_ATTR))
+            .andExpect(model().attributeExists(TEMPLATE_NAME_MODEL_ATTR))
+            .andExpect(model().attributeExists(TURNOVER_POLICY_MODEL_ATTR));
+    }
+
```
Comment (leahtaylor) [inline src/test/java/uk/gov/companieshouse/web/accounts/controller/smallfull/TurnoverPolicyControllerTest.java]: white space

File: src/test/java/uk/gov/companieshouse/web/accounts/service/smallfull/impl/TurnoverPolicyServiceImplTest.java [hunk 8]
```diff
@@ -30,6 +33,10 @@
     private static final String TRANSACTION_ID = "transactionId";
     private static final String COMPANY_ACCOUNTS_ID = "companyAccountsId";
```
Comment (leahtaylor) [inline src/test/java/uk/gov/companieshouse/web/accounts/service/smallfull/impl/TurnoverPolicyServiceImplTest.java]: Could the static imports all be grouped together?

File: src/test/java/uk/gov/companieshouse/web/accounts/service/smallfull/impl/TurnoverPolicyServiceImplTest.java [hunk 9]
```diff
@@ -30,6 +33,10 @@
...
 
+    private static final String TURNOVER_POLICY_DETAILS_FIELD_PATH = "turnoverPolicyDetails";
+    private static final String INVALID_STRING_SIZE_ERROR_MESSAGE = "validation.length.minInvalid.accounting_policies.turnover_policy";
+
+
```
Comment (leahtaylor) [inline src/test/java/uk/gov/companieshouse/web/accounts/service/smallfull/impl/TurnoverPolicyServiceImplTest.java]: whitespace

File: src/test/java/uk/gov/companieshouse/web/accounts/service/smallfull/impl/TurnoverPolicyServiceImplTest.java [hunk 10]
```diff
@@ -87,17 +95,42 @@ void shouldPostTurnoverPolicy() throws ServiceException {
         verify(accountingPoliciesServiceMock, times(1))
             .getAccountingPoliciesApi(TRANSACTION_ID, COMPANY_ACCOUNTS_ID);
 
-        verify(accountingPoliciesServiceMock, times(1))
-            .updateAccountingPoliciesApi(TRANSACTION_ID, COMPANY_ACCOUNTS_ID,
-                accountingPoliciesApiMock);
```
Comment (leahtaylor) [inline src/test/java/uk/gov/companieshouse/web/accounts/service/smallfull/impl/TurnoverPolicyServiceImplTest.java]: whitespace

File: src/main/resources/messages.properties (hunk 4 above)
Comment (ch-mcabrera) [reply src/main/resources/messages.properties]: `how` shouldn't be there

File: src/main/java/uk/gov/companieshouse/web/accounts/controller/smallfull/TurnoverPolicyController.java [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+@PreviousController(BasisOfPreparationController.class)
+@RequestMapping("/company/{companyNumber}/transaction/{transactionId}/company-accounts/{companyAccountsId}/small-full/turnover-policy")
+public class TurnoverPolicyController extends BaseController {
+
+    @Autowired
+    private TurnoverPolicyService turnoverPolicyService;
+
+    @GetMapping
+    public String getTurnoverPolicy(@PathVariable String companyNumber,
+        @PathVariable String transactionId,
```
Comment (BSaunders95) [inline src/main/java/uk/gov/companieshouse/web/accounts/controller/smallfull/TurnoverPolicyController.java]: Not important, but can we align the parameters of this method so the `@PathVariable`'s sit under one another? Makes for easier readability.

File: src/main/java/uk/gov/companieshouse/web/accounts/controller/smallfull/TurnoverPolicyController.java [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+
+        addBackPageAttributeToModel(model, companyNumber, transactionId, companyAccountsId);
+
+        if (bindingResult.hasErrors()) {
+            return getTemplateName();
+        }
+
+        try {
+            List<ValidationError> validationErrors = turnoverPolicyService
+                .postTurnoverPolicy(transactionId, companyAccountsId, turnoverPolicy);
```
Comment (BSaunders95) [inline src/main/java/uk/gov/companieshouse/web/accounts/controller/smallfull/TurnoverPolicyController.java]: Can we rename this method `submitTurnoverPolicy`? The controller needn't know about the type of `CRUD` operation performed by the service, plus since validation is now performed in this method 'submission' feels like it covers the functionality of the method

File: src/main/java/uk/gov/companieshouse/web/accounts/service/smallfull/impl/TurnoverPolicyServiceImpl.java [hunk 13]
```diff
@@ -47,6 +57,39 @@ public TurnoverPolicy getTurnOverPolicy(String transactionId, String companyAcco
...
+    }
 
+    /**
+     * Check if field contains invalid information when it is not null: empty or contain empty
+     * characters.
+     *
+     * @param turnoverPolicyDetails the string field that needs to be assessed
+     * @return
+     */
+    private boolean IsRequiredFieldEmpty(String turnoverPolicyDetails) {
```
Comment (BSaunders95) [inline src/main/java/uk/gov/companieshouse/web/accounts/service/smallfull/impl/TurnoverPolicyServiceImpl.java]: This method should begin with a lower case 'i'. Plus, is it worth grouping the `turnoverPolicy.getIsIncludeTurnoverSelected()` check in this method? Alternatively, just remove this method and bundle the empty check in the `if` check above.

File: src/main/java/uk/gov/companieshouse/web/accounts/service/smallfull/impl/TurnoverPolicyServiceImpl.java [hunk 14]
```diff
@@ -47,6 +57,39 @@ public TurnoverPolicy getTurnOverPolicy(String transactionId, String companyAcco
...
+        return turnoverPolicyDetails != null && StringUtils.trim(turnoverPolicyDetails).isEmpty();
```
Comment (BSaunders95) [inline src/main/java/uk/gov/companieshouse/web/accounts/service/smallfull/impl/TurnoverPolicyServiceImpl.java]: Could use `StringUtils.isBlank` here

File: src/main/resources/messages.properties [hunk 15]
```diff
@@ -74,4 +74,9 @@ validation.date.invalid.approval.date = The date cannot be on or before the end
 
 validation.characters.invalid.accounting_policies.basis_of_measurement_and_preparation = Enter valid characters
 validation.element.missing.accounting_policies.basis_of_measurement_and_preparation = Give details of how your accounts were prepared
-validation.length.maxExceeded.accounting_policies.basis_of_measurement_and_preparation = You must not exceed {0} characters
\ No newline at end of file
+validation.length.maxExceeded.accounting_policies.basis_of_measurement_and_preparation = You must not exceed {0} characters
+validation.length.maxExceeded.accounting_policies.basis_of_measurement_and_preparation = Basis of preparation must not exceed {0} characters
```
Comment (BSaunders95) [inline src/main/resources/messages.properties]: I assume this is a merge issue, but this string can be removed because the one above replaced it

File: src/main/resources/templates/fragments/inlineYesNoRadioButtons.html [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+        th:id="${radioButtonsFieldName} + '-errorId'"
+        th:if="${#fields.hasErrors('__${radioButtonsFieldName}__')}"
+        th:each="e : ${#fields.errors('__${radioButtonsFieldName}__')}" th:text="${e}">
+            </span>
+
+  <!-- Radio Buttons -->
+  <div class="govuk-radios govuk-radios--inline" data-module="radios">
+
+    <div class="govuk-radios__item">
+      <input class="govuk-radios__input show-text"
```
Comment (BSaunders95) [inline src/main/resources/templates/fragments/inlineYesNoRadioButtons.html]: Is there any time where the 'No' selection would reveal the text box? I assume not, but is it worth catering for the scenario by making the class dynamic. This could be done using `th:classappend`

File: src/test/java/uk/gov/companieshouse/web/accounts/controller/smallfull/TurnoverPolicyControllerTest.java [hunk 17]
```diff
@@ -0,0 +1,185 @@
...
+     * Add the information to the post request. Where the parameter will control if data to be set
+     * needs to be valid or invalid
+     *
+     * @param addInvalidData
+     * @return
+     */
+    private MockHttpServletRequestBuilder createPostRequestWithParam(boolean addInvalidData) {
+
+        String beanElement = "isIncludeTurnoverSelected";
+        String validData = addInvalidData ? null : "1";
```
Comment (BSaunders95) [inline src/test/java/uk/gov/companieshouse/web/accounts/controller/smallfull/TurnoverPolicyControllerTest.java]: Should we rename this parameter since it may not be 'valid' data?

File: src/test/java/uk/gov/companieshouse/web/accounts/service/smallfull/impl/TurnoverPolicyServiceImplTest.java [hunk 18]
```diff
@@ -86,18 +93,41 @@ void shouldPostTurnoverPolicy() throws ServiceException {
...
+        assertFalse(validationErrors.isEmpty());
+        assertEquals(validationErrors.get(0).getFieldPath(), TURNOVER_POLICY_DETAILS_FIELD_PATH);
+        assertEquals(validationErrors.get(0).getMessageKey(), INVALID_STRING_SIZE_ERROR_MESSAGE);
+    }
+
+    @Test
+    @DisplayName("Post a turnoverPolicy fails as turnoverPolicyDetails contains empty characters")
+    void shouldFailAsTurnoverPolicyDetailsContainsEmptyCharacters() throws ServiceException {
+
+        TurnoverPolicy turnoverPolicy = createTurnOverPolicy("    ");
```
Comment (BSaunders95) [inline src/test/java/uk/gov/companieshouse/web/accounts/service/smallfull/impl/TurnoverPolicyServiceImplTest.java]: Can we add another test which satisfies the `null` check?

File: start.sh [hunk 19]
```diff
@@ -27,4 +27,7 @@ else
     source "${APP_DIR}/app_env"
 fi
 
-exec java ${JAVA_MEM_ARGS} -jar -Dserver.port="${PORT}" "${APP_DIR}/company-accounts.web.ch.gov.uk.jar"
\ No newline at end of file
+exec java ${JAVA_MEM_ARGS} -jar -Dserver.port="${PORT}" "${APP_DIR}/company-accounts.web.ch.gov.uk.jar"
+
+# debug settings
```
Comment (BSaunders95) [inline start.sh]: Please can we revert this change?

File: src/main/java/uk/gov/companieshouse/web/accounts/controller/smallfull/TurnoverPolicyController.java (hunk 11 above)
Comment (ch-mcabrera) [reply src/main/java/uk/gov/companieshouse/web/accounts/controller/smallfull/TurnoverPolicyController.java]: I've used the intellij formatting which follows our coding.

File: src/main/java/uk/gov/companieshouse/web/accounts/controller/smallfull/TurnoverPolicyController.java (hunk 12 above)
Comment (ch-mcabrera) [reply src/main/java/uk/gov/companieshouse/web/accounts/controller/smallfull/TurnoverPolicyController.java]: Amost all controllers are following this naming convention. I dont think this change is really needed. Please, let me know what you think.

File: src/main/java/uk/gov/companieshouse/web/accounts/service/smallfull/impl/TurnoverPolicyServiceImpl.java (hunk 14 above)
Comment (ch-mcabrera) [reply src/main/java/uk/gov/companieshouse/web/accounts/service/smallfull/impl/TurnoverPolicyServiceImpl.java]: `StringUtils.isBlank` will return true when the string is null; which is invalid in my scenario.

File: src/main/resources/messages.properties (hunk 15 above)
Comment (ch-mcabrera) [reply src/main/resources/messages.properties]: it was.

File: start.sh (hunk 19 above)
Comment (ch-mcabrera) [reply start.sh]: i will do it. We should revert it in the company-accounts api to be consistent

File: src/main/resources/templates/fragments/inlineYesNoRadioButtons.html (hunk 16 above)
Comment (ch-mcabrera) [reply src/main/resources/templates/fragments/inlineYesNoRadioButtons.html]: After speaking with Brad, we both agreed we will do this change in the future if this is needed.

This fragment is used by accounting policies screens and it will show a text box when `yes` is selected.  we will revisit this If this logic changes(e.g. `no` displays the text);

File: src/main/java/uk/gov/companieshouse/web/accounts/controller/smallfull/TurnoverPolicyController.java (hunk 12 above)
Comment (ch-mcabrera) [reply src/main/java/uk/gov/companieshouse/web/accounts/controller/smallfull/TurnoverPolicyController.java]: After speaking with Brad. We agreed not to rename this method.

File: src/main/java/uk/gov/companieshouse/web/accounts/service/smallfull/impl/TurnoverPolicyServiceImpl.java (hunk 13 above)
Comment (ch-mcabrera) [reply src/main/java/uk/gov/companieshouse/web/accounts/service/smallfull/impl/TurnoverPolicyServiceImpl.java]: i moved this `turnoverPolicy.getIsIncludeTurnoverSelected()` to the method and rename method to make it more descriptive
